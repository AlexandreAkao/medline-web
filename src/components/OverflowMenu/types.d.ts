type OverflowMenuPosition = 'right' | 'left' | 'top' | 'bottom';

interface IOverflowMenuPosition {
  positionX: OverflowMenuPosition;
  positionY: OverflowMenuPosition;
}

interface IOverflowMenuProps {
  render: React.ReactNode;
  children: React.ReactNode;
}
