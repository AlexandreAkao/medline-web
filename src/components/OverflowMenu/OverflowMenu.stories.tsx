import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { FaBoxOpen } from 'react-icons/fa';

import OverflowMenu from 'components/OverflowMenu';

export default {
  title: 'OverflowMenu',
  component: OverflowMenu,
} as ComponentMeta<typeof OverflowMenu>;

const Template: ComponentStory<typeof OverflowMenu> = args => <OverflowMenu {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  render: 'Primary',
  children: (
    <div>
      <h1>Title</h1>
      <span>Option 1</span>
      <span>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Ex perferendis ipsum voluptates non nisi nihil
        accusantium, amet cupiditate molestiae qui aliquid dolores, numquam eligendi! Commodi recusandae impedit
        accusamus labore dolor.
      </span>
    </div>
  ),
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  render: <FaBoxOpen size={30} role="img" />,
  children: (
    <div>
      <h1>Title</h1>
      <span>Option 1</span>
      <button type="button" onClick={action('clicked')}>
        Click me
      </button>
    </div>
  ),
};
