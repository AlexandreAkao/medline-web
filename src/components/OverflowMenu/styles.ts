import styled from 'styled-components';

import colors from 'styles/colors';
import zIndexPriority from 'styles/zIndexPriority';

export const OverflowMenuContainer = styled.div`
  display: inline;
  position: relative;
  cursor: pointer;
`;

export const OverflowMenuContent = styled.div<IOverflowMenuPosition>`
  z-index: ${zIndexPriority.ALMOST_MAX};
  position: absolute;
  background-color: ${colors.white.normal};
  color: black;
  display: flex;
  flex-direction: column;
  box-shadow: 0 4px 8px rgb(0 0 0 / 10%);
  border: 1px solid ${colors.white.lighter};
  border-radius: 4px;
  width: clamp(200px, 300px, 350px);

  ${({ positionX, positionY }) => ({
    [positionX]: 0,
    [positionY]: 0,
  })}
`;
