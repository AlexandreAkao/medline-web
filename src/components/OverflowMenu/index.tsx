import { MouseEvent, useRef, useState } from 'react';

import { OverflowMenuContainer, OverflowMenuContent } from 'components/OverflowMenu/styles';
import useOutsideHandler from 'hooks/useOutsideHandler';

function OverflowMenu({ children, render }: IOverflowMenuProps) {
  const [isOpen, setIsOpen] = useState(false);
  const [position, setPosition] = useState<IOverflowMenuPosition>({
    positionX: 'right',
    positionY: 'top',
  });
  const ref = useRef<HTMLDivElement>(null);
  useOutsideHandler<HTMLDivElement>(ref, () => setIsOpen(false));

  const handleClick = (event: MouseEvent<HTMLDivElement>) => {
    const { clientX, clientY, view } = event;
    const clientHeight = view?.document?.body?.clientHeight ?? 100;
    const popupHeight = ref.current?.children[0]?.clientHeight ?? 100;
    const positionX = clientX < 350 ? 'left' : 'right';
    const positionY = clientHeight - clientY < popupHeight ? 'bottom' : 'top';

    setPosition({ positionX, positionY });
    setIsOpen(true);
  };

  return (
    <OverflowMenuContainer ref={ref} onClick={handleClick}>
      {render}
      {isOpen && (
        <OverflowMenuContent data-testid="overflow-menu-content" {...position}>
          {children}
        </OverflowMenuContent>
      )}
    </OverflowMenuContainer>
  );
}

export default OverflowMenu;
