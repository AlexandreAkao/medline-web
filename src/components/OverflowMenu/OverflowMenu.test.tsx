import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';

import * as stories from 'components/OverflowMenu/OverflowMenu.stories';
import OverflowMenu from 'components/OverflowMenu';

const { Primary, WithIcon } = composeStories(stories);

describe('Checkbox', () => {
  const renderComponent = (customProps: Partial<IOverflowMenuProps> = {}) => {
    const props: IOverflowMenuProps = {
      render: 'Example',
      children: undefined,
      ...customProps,
    };
    return render(<OverflowMenu {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render a primary Button', () => {
    render(<Primary />);
    const labelElement = screen.getByText(/Primary/i);

    expect(labelElement).toBeInTheDocument();
  });

  it('should render a primary Button', () => {
    render(<WithIcon />);
    const iconElement = screen.getByRole('img');

    expect(iconElement).toBeInTheDocument();
  });

  it('should open overflow on click', async () => {
    renderComponent();
    const labelElement = screen.getByText(/Example/i);

    await userEvent.click(labelElement);

    const overflowElement = screen.getByTestId('overflow-menu-content');
    expect(labelElement).toBeInTheDocument();
    expect(overflowElement).toBeInTheDocument();
  });
});
