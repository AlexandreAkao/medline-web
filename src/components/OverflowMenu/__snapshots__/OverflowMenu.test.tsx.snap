// Vitest Snapshot v1

exports[`Checkbox > should match with snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ fqRZws"
  >
    Example
  </div>
</div>
`;
